sim/xcelium/__init__.py

```diff
@@ -0,0 +1,394 @@
+from hammer_vlsi import FlowLevel, TimeValue
+from hammer_vlsi import HammerSimTool, HammerToolStep, HammerLSFSubmitCommand, HammerLSFSettings
+from hammer_logging import HammerVLSILogging
+
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)),"../../common"))
+from tool import CadenceTool
+
+class xcelium(HammerSimTool, CadenceTool):
+
+  @property
+  def xcelium_ext(self) -> List[str]:
+    verilog_ext  = [".v", ".V", ".VS", ".vp", ".VP"]
+    sverilog_ext = [".sv",".SV",".svp",".SVP",".svi",".svh",".vlib",".VLIB"]
+    c_cxx_ext    = [".c",".cc",".cpp"]
+    gz_ext       = [ext + ".gz" for ext in verilog_ext + sverilog_ext]
+    z_ext        = [ext + ".z" for ext  in verilog_ext + sverilog_ext]
+    return (verilog_ext + sverilog_ext + c_cxx_ext + gz_ext + z_ext)
+
+  @property
+  def steps(self) -> List[HammerToolStep]:
+    return self.make_steps_from_methods([self.compile_xrun,
+                                         self.elaborate_xrun,
+                                         self.sim_xrun])
+    
+  @property
+  def tool_config_prefix(self) -> str:
+    return "sim.xcelium"
+  
+  @property
+  def sim_input_prefix(self) -> str:
+    return "sim.inputs"
+  
+  @property
+  def sim_waveform_prefix(self) -> str:
+    return "sim.inputs.waveform"
+  
+  @property
+  def xcelium_bin(self) -> str:
+    return self.get_setting("sim.xcelium.xcelium_bin")
+
+  @property
+  def sim_tcl_file(self) -> str: 
+    return os.path.join(self.run_dir, "xrun_sim.tcl")
+    
+  @property
+  def sdf_cmd_file(self) -> str:
+    return os.path.join(self.run_dir, "design.sdf_cmd")
+  
+  def post_synth_sdc(self) -> Optional[str]:
+    pass
+
+  def write_gl_files(self) -> bool:
+    return True
+  
+  def get_verilog_models(self) -> List[str]:
+    verilog_sim_files = self.technology.read_libs([
+        hammer_tech.filters.verilog_sim_filter], 
+        hammer_tech.HammerTechnologyUtils.to_plain_item)
+    return verilog_sim_files
+        
+  def fill_outputs(self) -> bool:
+    self.output_waveforms = []
+    self.output_saifs = []
+    self.output_top_module = self.top_module
+    self.output_tb_name = self.get_setting(f"{self.sim_input_prefix}.tb_name")
+    self.output_tb_dut = self.get_setting(f"{self.sim_input_prefix}.tb_dut")
+    self.output_level = self.get_setting(f"{self.sim_input_prefix}.level")
+    return True
+   
+  # Several extract functions are used to process mandatory keys into string options. 
+  # Returns a raw input dictionary as well.
+
+  def extract_xrun_opts(self) -> Tuple[Dict[str, str], Dict[str, str]]:
+    xrun_opts_def = [("enhanced_recompile", True),
+                     ("xmlibdirname", None),
+                     ("xmlibdirpath", None),
+                     ("simtmp", None),
+                     ("snapshot", None),
+                     ("global_access", False)]
+
+    xrun_opts = {opt[0] : self.get_setting(f"{self.tool_config_prefix}.{opt[0]}", opt[1]) for opt in xrun_opts_def}
```

Comment:
I quite like this `get_setting` method - would be nice if was made more global

Reply:
Perhaps could become a convenience function in hammer_utils?

Reply:
I'd put it in `hammer/config/config_src.py` and have `HammerTool` inherit from it.

Reply:
Will PR a new function `get_setting_from_list` in main hammer.